File: app/src/main/java/me/ketal/hook/LeftSwipeReplyHook.java
package me.ketal.hook;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Looper;
import android.view.View;
import android.widget.ImageView;
import android.widget.Toast;

import java.lang.reflect.Method;

import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XposedBridge;
import de.robv.android.xposed.XposedHelpers;
import me.singleneuron.qn_kernel.tlb.ConfigTable;
import nil.nadph.qnotified.SyncUtils;
import nil.nadph.qnotified.config.ConfigManager;
import nil.nadph.qnotified.hook.CommonDelayableHook;
import nil.nadph.qnotified.step.DexDeobfStep;
import nil.nadph.qnotified.ui.ResUtils;
import nil.nadph.qnotified.util.DexKit;
import nil.nadph.qnotified.util.Toasts;
import nil.nadph.qnotified.util.Utils;

import static me.ketal.util.TIMVersion.TIM_3_1_1;
import static me.singleneuron.util.QQVersion.QQ_8_2_6;
import static nil.nadph.qnotified.util.Initiator._BaseChatPie;
import static nil.nadph.qnotified.util.Initiator._ChatMessage;
import static nil.nadph.qnotified.util.Utils.TOAST_TYPE_ERROR;
import static nil.nadph.qnotified.util.Utils.findMethodByTypes_1;
import static nil.nadph.qnotified.util.Utils.getApplication;
import static nil.nadph.qnotified.util.Utils.getFirstByType;
import static nil.nadph.qnotified.util.Utils.getHostVersionCode;
import static nil.nadph.qnotified.util.Utils.hasMethod;
import static nil.nadph.qnotified.util.Utils.invoke_virtual_any;
import static nil.nadph.qnotified.util.Utils.isTim;
import static nil.nadph.qnotified.util.Utils.log;

public class LeftSwipeReplyHook extends CommonDelayableHook {
    public static final LeftSwipeReplyHook INSTANCE = new LeftSwipeReplyHook();
    public static final String LEFT_SWIPE_NOACTION = "ketal_left_swipe_noAction";
    public static final String LEFT_SWIPE_MULTICHOOSE = "ketal_left_swipe_multiChoose";
    public static final String LEFT_SWIPE_REPLYDISTANCE = "ketal_left_swipe_replyDistance";
    public static final int FLAG_REPLACE_PIC = 10001;
    private static Bitmap img;

    protected LeftSwipeReplyHook() {
        super("ketal_left_swipe_action", SyncUtils.PROC_MAIN, true, new DexDeobfStep(DexKit.N_LeftSwipeReply_Helper__reply), new DexDeobfStep(DexKit.N_BASE_CHAT_PIE__chooseMsg));
    }

    private static Bitmap getMultiBitmap() {
        if (img == null || img.isRecycled())
            img = BitmapFactory.decodeStream(ResUtils.openAsset("list_checkbox_selected_nopress.png"));
        return img;
    }

    @Override
    public boolean isValid() {
        if (isTim() && getHostVersionCode() >= TIM_3_1_1)
            return true;
        else return !isTim() && getHostVersionCode() >= QQ_8_2_6;
    }

    @Override
    protected boolean initOnce() {
        try {
            Method replyMethod = DexKit.doFindMethod(DexKit.N_LeftSwipeReply_Helper__reply);
            Class hookClass = replyMethod.getDeclaringClass();
            String methodName = isTim() ? "L" : "a";
            XposedHelpers.findAndHookMethod(hookClass, methodName, float.class, float.class, new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                    if (!isEnabled())
                        return;
                    if (isNoAction())
                        param.setResult(null);
                }
            });

            XposedBridge.hookMethod(findMethodByTypes_1(hookClass, void.class, View.class, int.class), new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                    if (!isEnabled() || !isMultiChose())
                        return;
                    ImageView iv = (ImageView) param.args[0];
                    if (iv.getTag(FLAG_REPLACE_PIC) == null) {
                        iv.setImageBitmap(getMultiBitmap());
                        iv.setTag(FLAG_REPLACE_PIC, true);
                    }
                }
            });

            XposedBridge.hookMethod(replyMethod, new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                    if (!isEnabled() || !isMultiChose())
                        return;
                    Object message = invoke_virtual_any(param.thisObject, _ChatMessage());
                    Object baseChatPie = getFirstByType(param.thisObject, _BaseChatPie());
                    DexKit.doFindMethod(DexKit.N_BASE_CHAT_PIE__chooseMsg).invoke(baseChatPie, message);
                    param.setResult(null);
                }
            });

            methodName = isTim() ? ConfigTable.INSTANCE.getConfig(LeftSwipeReplyHook.class.getSimpleName()) : "a";
            XposedBridge.hookMethod(hasMethod(hookClass, methodName, int.class), new XC_MethodHook() {
                @Override
                protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                    if (!isEnabled())
                        return;
                    if (getReplyDistance() <= 0)
                        setReplyDistance((Integer) param.getResult());
                    else
                        param.setResult(getReplyDistance());
                }
            });
            return true;
        } catch (Exception e) {
            log(e);
        }
        return false;
    }

    public boolean isNoAction() {
        return ConfigManager.getDefaultConfig().getBooleanOrDefault(LEFT_SWIPE_NOACTION, false);
    }

    public void setNoAction(boolean on) {
        putValue(LEFT_SWIPE_NOACTION, on);
    }

    public boolean isMultiChose() {
        return ConfigManager.getDefaultConfig().getBooleanOrDefault(LEFT_SWIPE_MULTICHOOSE, false);
    }

    public void setMultiChose(boolean on) {
        putValue(LEFT_SWIPE_MULTICHOOSE, on);
    }

    public int getReplyDistance() {
        return (int) ConfigManager.getDefaultConfig().getOrDefault(LEFT_SWIPE_REPLYDISTANCE, -1);
    }

    public void setReplyDistance(int replyDistance) {
        putValue(LEFT_SWIPE_REPLYDISTANCE, replyDistance);
    }

    private void putValue(String keyName, Object object) {
        try {
            ConfigManager mgr = ConfigManager.getDefaultConfig();
            mgr.getAllConfig().put(keyName, object);
            mgr.save();
        } catch (Exception e) {
            Utils.log(e);
            if (Looper.myLooper() == Looper.getMainLooper()) {
                Toasts.showToast(getApplication(), TOAST_TYPE_ERROR, e + "", Toast.LENGTH_SHORT);
            } else {
                SyncUtils.post(new Runnable() {
                    @Override
                    public void run() {
                        Toasts.showToast(getApplication(), TOAST_TYPE_ERROR, e + "", Toast.LENGTH_SHORT);
                    }
                });
            }
        }
    }
}

File: app/src/main/java/me/ketal/hook/LeftSwipeReplyHook.kt
/*
 * QNotified - An Xposed module for QQ/TIM
 * Copyright (C) 2019-2021 dmca@ioctl.cc
 * https://github.com/ferredoxin/QNotified
 *
 * This software is non-free but opensource software: you can redistribute it
 * and/or modify it under the terms of the GNU Affero General Public License
 * as published by the Free Software Foundation; either
 * version 3 of the License, or any later version and our eula as published
 * by ferredoxin.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * and eula along with this software.  If not, see
 * <https://www.gnu.org/licenses/>
 * <https://github.com/ferredoxin/QNotified/blob/master/LICENSE.md>.
 */
package me.ketal.hook

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.view.View
import android.widget.ImageView
import de.robv.android.xposed.XposedHelpers
import ltd.nextalone.util.hookAfter
import ltd.nextalone.util.hookBefore
import me.ketal.data.ConfigData
import me.ketal.util.BaseUtil.tryVerbosely
import me.ketal.util.PlayQQVersion
import me.ketal.util.TIMVersion
import me.singleneuron.qn_kernel.data.hostInfo
import me.singleneuron.qn_kernel.data.requireMinVersion
import me.singleneuron.qn_kernel.tlb.ConfigTable.getConfig
import me.singleneuron.util.QQVersion
import nil.nadph.qnotified.base.annotation.FunctionEntry
import nil.nadph.qnotified.hook.CommonDelayableHook
import nil.nadph.qnotified.step.DexDeobfStep
import nil.nadph.qnotified.ui.ResUtils
import nil.nadph.qnotified.util.*

@FunctionEntry
object LeftSwipeReplyHook : CommonDelayableHook(
    "ketal_left_swipe_action",
    DexDeobfStep(DexKit.N_LeftSwipeReply_Helper__reply),
    DexDeobfStep(DexKit.N_BASE_CHAT_PIE__chooseMsg)
) {
    private val LEFT_SWIPE_NO_ACTION = ConfigData<Boolean>("ketal_left_swipe_noAction")
    private val LEFT_SWIPE_MULTI_CHOOSE = ConfigData<Boolean>("ketal_left_swipe_multiChoose")
    private val LEFT_SWIPE_REPLY_DISTANCE = ConfigData<Int>("ketal_left_swipe_replyDistance")
    var isNoAction: Boolean
        get() = LEFT_SWIPE_NO_ACTION.getOrDefault(false)
        set(on) {
            LEFT_SWIPE_NO_ACTION.value = on
        }
    var isMultiChose: Boolean
        get() = LEFT_SWIPE_MULTI_CHOOSE.getOrDefault(false)
        set(on) {
            LEFT_SWIPE_MULTI_CHOOSE.value = on
        }
    var replyDistance: Int
        get() = LEFT_SWIPE_REPLY_DISTANCE.getOrDefault(-1)
        set(replyDistance) {
            LEFT_SWIPE_REPLY_DISTANCE.value = replyDistance
        }
    private var img: Bitmap? = null
    private val multiBitmap: Bitmap?
        get() {
            if (img == null || img!!.isRecycled) img =
                BitmapFactory.decodeStream(ResUtils.openAsset("list_checkbox_selected_nopress.png"))
            return img
        }

    override fun isValid(): Boolean = requireMinVersion(QQVersion.QQ_8_2_6, TIMVersion.TIM_3_1_1, PlayQQVersion.PlayQQ_8_2_9)

    override fun initOnce() = tryVerbosely(false) {
        val replyMethod = DexKit.doFindMethod(DexKit.N_LeftSwipeReply_Helper__reply)
        val hookClass = replyMethod!!.declaringClass
        var methodName = if (hostInfo.isTim) "L" else "a"
        XposedHelpers.findMethodBestMatch(
            hookClass,
            methodName,
            Float::class.java,
            Float::class.java
        )
            .hookBefore(this) {
                if (isNoAction) it.result = null
            }
        ReflexUtil.findMethodByTypes_1(
            hookClass,
            Void.TYPE,
            View::class.java,
            Int::class.javaPrimitiveType
        )
            .hookBefore(this) {
                if (!isMultiChose) return@hookBefore
                val iv = it.args[0] as ImageView
                if (iv.tag == null) {
                    iv.setImageBitmap(multiBitmap)
                    iv.tag = true
                }
            }
        replyMethod.hookBefore(this) {
            if (!isMultiChose) return@hookBefore
            val message = ReflexUtil.invoke_virtual_any(it.thisObject, Initiator._ChatMessage())
            val baseChatPie =
                ReflexUtil.getFirstByType(it.thisObject, Initiator._BaseChatPie() as Class<*>)
            DexKit.doFindMethod(DexKit.N_BASE_CHAT_PIE__chooseMsg)!!.invoke(baseChatPie, message)
            it.result = null
        }
        methodName =
            if (hostInfo.isTim) getConfig(LeftSwipeReplyHook::class.java.simpleName) else "a"
        ReflexUtil.hasMethod(hookClass, methodName, Int::class.java)
            .hookAfter(this) {
                if (replyDistance <= 0) {
                    replyDistance = it.result as Int
                } else {
                    it.result = replyDistance
                }
            }
        true
    }
}
